service/openapi.yaml

```diff
@@ -0,0 +1,285 @@
+    post:
+      description: OFREP flag evaluation request
+      parameters:
+        - name: key
+          in: path
+          required: true
+          schema:
+            type: string
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/evaluationRequest'
+      responses:
+        '200':
+          description: OFREP successful evaluation response
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/success'
+        '400':
+          description: OFREP error response
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/failure'
+        '401':
+          description: Unauthorized - You need credentials to access the API
+        '403':
+          description: Forbidden - You are not authorized to access the API
+        '404':
+          description: Flag not found
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/failure'
+        '429':
+          description: Rate limit reached on the Flag Management System
+          headers:
+            Retry-Later:
+              description: Indicates when to retry the request
+              schema:
+                type: string
+                format: date-time
+                example: '2024-02-07T12:00:00Z'
+        '500':
+          description: Internal server error
+  /ofrep/v1/evaluate:
+    post:
+      description: OFREP bulk evaluation request
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/bulkEvaluationRequest'
+      responses:
+        '200':
+          description: OFREP successful evaluation response
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/bulkEvaluationSuccess'
+        '400':
+          description: OFREP error response
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/bulkEvaluationFailure'
+        '401':
+          description: Unauthorized - You need credentials to access the API
+        '403':
+          description: Forbidden - You are not authorized to access the API
+        '429':
+          description: Rate limit reached on the Flag Management System
+          headers:
+            Retry-Later:
+              description: Indicates when to retry the request
+              schema:
+                type: string
+                format: date-time
+                example: '2024-02-07T12:00:00Z'
+        '500':
+          description: Internal server error
+  /ofrep/v1/flag/changes:
+    post:
+      description: OFREP polling endpoint to check if the flags have been modified
+      requestBody:
+        content:
+          application/json:
+            schema:
```

Comment:
Not 100% sure how this endpoint will work. Let me know if I've understood correctly:

The provider can use this endpoint to understand if the flag "rules" in the flag provider system (backend) have changed since the providers last evaluation and decide to make an evaluate call even if the evaluation context has not changed?

Reply:
Yes sorry if the context is not clear, but this is exactly that.

The goal of this endpoint is to do polling on it if the `ETag` is different than the one in the cache of the provider we will call the `evaluate` endpoint for the changed flags.

It avoids having to go through re-evaluation too often.
**Why are we doing polling?** Because this is the most voted one in the OFREP survey, but in the future we should also probably support Server Send Event or Websocket

Reply:
If you want to have a look I've started a very draft PR of an OFREP web provider https://github.com/open-feature/js-sdk-contrib/pull/776.

It may help you to see how we use this endpoint.

Reply:
Why does this need to be a different endpoint? Couldn't you set the `ETag` header on any of the other endpoints and have them return `304` responses? 

I guess it depends on how the provider is implementing `ETag` headers for these endpoints, but I assume they would be a hash on the FF configuration for a user, and not the hash of the underlying project configuration. Because in most platforms flag evaluation can be dynamic based on things like schedules that wouldn't be captured by the project configuration `ETag`.

Reply:
The main goal of having a different endpoint was to avoid having to re-evaluate all the flags, but your point about evaluation being dynamic is super interesting.

I am not opposed to regularly re-evaluating and sending the ETag back _(just to ease the comparison)_,  but this can put pressure on the different flag management systems.

About the 304, how do you imagine doing it on the server side? Because it does not have the context of the provider, do you expect the flag management systems to keep some state?

Reply:
I would like to avoid adding a different endpoint if possible. As @jonathannorris mentioned, we should be able to reuse the same API and return a `304` if the ruleset hasn't changed or updated context hasn't affected the evaluated response. This would require context to be sent in every request but the overhead should be minimal.

Reply:
Do we really want to re-evaluate everything all the time?

Reply:
I kind of question the value of the `ETag` header here at all; for most providers with some dynamic evaluation, we'd likely be doing a complete evaluation every time anyways. Therefore, the only thing we'd be saving is the bandwidth of returning the flag results in a `200` response vs a `304` response. 

Maybe letting providers set `max-age` headers in their response and using standard caching headers from there may let providers control how often they want to re-evaluate flags. But from DevCycle's perspective, we'd just re-evaluate every time; we have a dynamic evaluation at the edge where we can merge client-side contexts with our edgeDB user database, in addition to schedules and other rollout mechanics that should be evaluated every time. Also, a user evaluation is usually less than the overhead of accepting the HTTP request.

Reply:
> Also, a user evaluation is usually less than the overhead of accepting the HTTP request.

I suspect this is generally the case. I think opening a TCP connection, securing it, and  deserializing a JSON payload is probably more work than evaluating tens of flags in most systems.

> I kind of question the value of the ETag header here at all; for most providers with some dynamic evaluation, we'd likely be doing a complete evaluation every time anyways. Therefore, the only thing we'd be saving is the bandwidth of returning the flag results in a 200 response vs a 304 response.

If you have context-insensitive flags (a simple rollout flag that's on/off for everyone, for example) and `ETag`/`If-None-Match` could be valuable; in that case if the provider sends `If-None-Match: abc123` and that matches the currrent `ETag`, we know we can actually skip evaluation.

> Maybe letting providers set max-age headers in their response and using standard caching headers from there may let providers control how often they want to re-evaluate flags.

I like this idea.

I think overall we can get the optimal caching behavior we want without the dedicated `/change` endpoint. For an implementation to determine if it should send a 304, it would check the `If-None-Match` header and also determine if any of the flags implicated in the request are context sensitive. Additionally, if no context is included in the request, it might be safe for an implementation to assume it can send a 304.

Reply:
I am fine to go for re-evaluation if we think this is the way to go.

So I try to sum up the discussion here:
- We can get rid of the `/change` endpoint.
- It is fine to ask for re-evaluation each time.
- We will have an `ETag` as a header that is computed for the full request _(not per flag)_.

Is this what you had in mind @toddbaert @jonathannorris ?

Reply:
I think that's the right direction for an initial version.